### PR TITLE
Update Array builtins with es2015 code path

### PR DESCRIPTION
**Pull request by pmarkee**
Related issue: #191

Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 1]
```diff
@@ -1015,8 +1061,15 @@ static Value builtinArrayMap(ExecutionState& state, Value thisValue, size_t argc
     if (argc > 1)
         T = argv[1];
 
+#ifdef ESCARGOT_ENABLE_ES2015
+    // Let A be ArraySpeciesCreate(O, len).
+    Value val = arraySpeciesCreate(state, O, len);
+    ASSERT(val.isObject() && val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: please separate ASSERT.
ASSERT(val.isObject() == true);
ASSERT(val.asObject()->isArrayObject()) == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 2]
```diff
@@ -953,8 +992,15 @@ static Value builtinArrayFilter(ExecutionState& state, Value thisValue, size_t a
     if (argc > 1)
         T = argv[1];
 
+#ifdef ESCARGOT_ENABLE_ES2015
+    // Let A be ArraySpeciesCreate(O, 0).
+    Value val = arraySpeciesCreate(state, O, 0);
+    ASSERT(val.isObject() && val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: please separate ASSERT.
ASSERT(val.isObject() == true);
ASSERT(val.asObject()->isArrayObject()) == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 3]
```diff
@@ -977,8 +1023,8 @@ static Value builtinArrayFilter(ExecutionState& state, Value thisValue, size_t a
 
             // If ToBoolean(selected) is true, then
             if (selected.toBoolean(state)) {
-                // Call the [[DefineOwnProperty]] internal method of A with arguments ToString(to), Property Descriptor {[[Value]]: kValue, [[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: true}, and false.
-                A->defineOwnProperty(state, ObjectPropertyName(state, Value(to)), ObjectPropertyDescriptor(kValue, ObjectPropertyDescriptor::AllPresent));
+                // Let status be CreateDataPropertyOrThrow (A, ToString(to), kValue).
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Check before dereferencing
ASSERT(A != nullptr);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 4]
```diff
@@ -1034,8 +1087,8 @@ static Value builtinArrayMap(ExecutionState& state, Value thisValue, size_t argc
             // Let mappedValue be the result of calling the [[Call]] internal method of callbackfn with T as the this value and argument list containing kValue, k, and O.
             Value v[] = { kValue, Value(k), O };
             Value mappedValue = callbackfn.asFunction()->call(state, T, 3, v);
-            // Call the [[DefineOwnProperty]] internal method of A with arguments Pk, Property Descriptor {[[Value]]: mappedValue, [[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: true}, and false.
-            A->defineOwnProperty(state, Pk, ObjectPropertyDescriptor(mappedValue, ObjectPropertyDescriptor::AllPresent));
+            // Let status be CreateDataPropertyOrThrow (A, Pk, mappedValue).
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Check before dereferencing
ASSERT(A != nullptr);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 5]
```diff
@@ -661,7 +692,15 @@ static Value builtinArraySlice(ExecutionState& state, Value thisValue, size_t ar
     uint32_t finalEnd = (relativeEnd < 0) ? std::max((double)len + relativeEnd, 0.0) : std::min(relativeEnd, (double)len);
 
     int64_t n = 0;
+#ifdef ESCARGOT_ENABLE_ES2015
+    // Let count be max(final - k, 0).
+    // Let A be ArraySpeciesCreate(O, count).
+    Value val = arraySpeciesCreate(state, thisObject, std::max(finalEnd - k, (int64_t)0));
+    ASSERT(val.isObject() && val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: please separate ASSERT.
ASSERT(val.isObject() == true);
ASSERT(val.asObject()->isArrayObject()) == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 6]
```diff
@@ -408,16 +451,47 @@ static Value builtinArraySplice(ExecutionState& state, Value thisValue, size_t a
...
+        // Let dc be ToInteger(deleteCount).
+        double dc = argv[1].toInteger(state);
+        // Let actualDeleteCount be min(max(dc,0), len – actualStart).
+        actualDeleteCount = std::min(std::max(dc, 0.0), (double)(len - actualStart));
+    }
+    // If len+insertCount−actualDeleteCount > 2^53-1, throw a TypeError exception.
+    CHECK_ARRAY_LENGTH(len + insertCount - actualDeleteCount, (1LL << 53));
+    // Let A be ArraySpeciesCreate(O, actualDeleteCount).
+    Value val = arraySpeciesCreate(state, O, actualDeleteCount);
+    ASSERT(val.isObject() && val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: please separate ASSERT.
ASSERT(val.isObject() == true);
ASSERT(val.asObject()->isArrayObject()) == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 7]
```diff
@@ -431,9 +505,8 @@ static Value builtinArraySplice(ExecutionState& state, Value thisValue, size_t a
         // Let fromValue be the result of calling the [[Get]] internal method of O with argument from.
         ObjectGetResult fromValue = O->getIndexedProperty(state, Value(actualStart + k));
         if (fromValue.hasValue()) {
-            // Call the [[DefineOwnProperty]] internal method of A with arguments ToString(k), Property Descriptor {[[Value]]: fromValue, [[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: true}, and false.
-            A->ArrayObject::defineOwnProperty(state, ObjectPropertyName(state, Value(k)),
-                                              ObjectPropertyDescriptor(fromValue.value(state, O), ObjectPropertyDescriptor::AllPresent));
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Check before dereferencing
ASSERT(A != nullptr);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 8]
```diff
@@ -76,6 +76,47 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+    if (!C.isConstructor()) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, state.context()->staticStrings().Array.string(), false, String::emptyString, errorMessage_GlobalObject_ThisNotConstructor);
+    }
+    // Return Construct(C, <<length>>).
+    Value argv[1] = { Value(length) };
+    return C.asFunction()->call(state, C, 1, argv);
+}
+
 static Value builtinArrayIsArray(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
 {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: ASSERT(argv != nullptr);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 9]
```diff
@@ -76,6 +76,47 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
     return array;
 }
 
+#define CHECK_ARRAY_LENGTH(num, limit)                                                                               \
+    if ((num) >= (limit)) {                                                                                          \
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, errorMessage_GlobalObject_InvalidArrayLength); \
+    }
+
+static Value arraySpeciesCreate(ExecutionState& state, Object* originalArray, const size_t length)
+{
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: Check pointer-type parameter / check before deferencing
ASSERT(originalArray != nullptr);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 10]
```diff
@@ -76,6 +76,47 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+    // Assert: length is an integer Number >= 0.
+    ASSERT((int64_t)length >= 0);
+    Value originalArrayConstructor = originalArray->get(state, ObjectPropertyName(state.context()->staticStrings().constructor)).value(state, originalArray);
+
+    // Let C be undefined.
+    Value C;
+    // Let isArray be IsArray(originalArray).
+    // If isArray is true, then
+    if (originalArray->isArrayObject()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: State expression explicityl
if (originalArray->isArrayObject() == true)

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 11]
```diff
@@ -76,6 +76,47 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+        // Let C be Get(originalArray, "constructor").
+        C = originalArrayConstructor;
+
+        // TODO 9.4.2.3. 6.c. (after Realm is implemented)
+        // If Type(C) is Object, then
+        if (C.isObject()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: if (C.isObject() == true)

**Comment by pmarkee** [comment]
@yichoi @clover2123 can you please take a look at this patch? Thank you.

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 12]
```diff
@@ -76,8 +76,51 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+        // If Type(C) is Object, then
+        if (C.isObject() == true) {
+            // Let C be Get(C, @@species).
+            C = C.asObject()->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().species)).value(state, C);
+        }
+    }
+
+    // If C is null, let C be undefined.
+    // If C is undefined, return ArrayCreate(length).
+    if (C.isUndefinedOrNull()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
if (C.isUndefinedOrNull() == true)

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 13]
```diff
@@ -76,8 +76,51 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+        return new ArrayObject(state, static_cast<double>(length));
+    }
+    // If IsConstructor(C) is false, throw a TypeError exception.
+    if (!C.isConstructor()) {
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
if (C.isConstructor() == false)

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 14]
```diff
@@ -408,16 +453,49 @@ static Value builtinArraySplice(ExecutionState& state, Value thisValue, size_t a
...
+        // Let dc be ToInteger(deleteCount).
+        double dc = argv[1].toInteger(state);
+        // Let actualDeleteCount be min(max(dc,0), len – actualStart).
+        actualDeleteCount = std::min(std::max(dc, 0.0), (double)(len - actualStart));
+    }
+    // If len+insertCount−actualDeleteCount > 2^53-1, throw a TypeError exception.
+    CHECK_ARRAY_LENGTH(len + insertCount - actualDeleteCount, (1LL << 53));
+    // Let A be ArraySpeciesCreate(O, actualDeleteCount).
+    Value val = arraySpeciesCreate(state, O, actualDeleteCount);
+    ASSERT(val.isObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
ASSERT(val.isObject() == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 15]
```diff
@@ -408,16 +453,49 @@ static Value builtinArraySplice(ExecutionState& state, Value thisValue, size_t a
...
+    ASSERT(val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
ASSERT(val.isObject()->isArrayObject());

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 16]
```diff
@@ -661,7 +697,16 @@ static Value builtinArraySlice(ExecutionState& state, Value thisValue, size_t ar
     uint32_t finalEnd = (relativeEnd < 0) ? std::max((double)len + relativeEnd, 0.0) : std::min(relativeEnd, (double)len);
 
     int64_t n = 0;
+#ifdef ESCARGOT_ENABLE_ES2015
+    // Let count be max(final - k, 0).
+    // Let A be ArraySpeciesCreate(O, count).
+    Value val = arraySpeciesCreate(state, thisObject, std::max(finalEnd - k, (int64_t)0));
+    ASSERT(val.isObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
ASSERT(val.isObject() == true);

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 17]
```diff
@@ -661,7 +697,16 @@ static Value builtinArraySlice(ExecutionState& state, Value thisValue, size_t ar
...
+    ASSERT(val.asObject()->isArrayObject());
```
Comment (yichoi) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For readability / State expression explicitly
ASSERT(val.isObject()->isArrayObject());

File: src/runtime/GlobalObjectBuiltinArray.cpp [hunk 18]
```diff
@@ -76,8 +76,51 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
...
+    if (C.isUndefinedOrNull() == true) {
+        return new ArrayObject(state, static_cast<double>(length));
+    }
+    // If IsConstructor(C) is false, throw a TypeError exception.
+    if (C.isConstructor() == false) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, state.context()->staticStrings().Array.string(), false, String::emptyString, errorMessage_GlobalObject_ThisNotConstructor);
+    }
+    // Return Construct(C, <<length>>).
+    Value argv[1] = { Value(length) };
+    return C.asFunction()->call(state, C, 1, argv);
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinArray.cpp]: For `Construct`, `ByteCodeInterpreter::newOperation()` is used instead.
(Our current implementation for `Construct` is imperfect, so I'm working on revising it)
And would you please add internal TCs for newly added Array builtins?

**Pull request merged**